kopykat-ksp/src/main/kotlin/at/kopyk/utils/TypeCompileScope.kt

```diff
@@ -55,42 +59,85 @@ internal sealed interface TypeCompileScope : KSClassDeclaration {
   fun toFileScope(file: FileSpec.Builder): FileCompilerScope
 }
 
+internal fun TypeParameterResolver.invariant() = object : TypeParameterResolver {
+  override val parametersMap: Map<String, TypeVariableName>
+    get() = this@invariant.parametersMap.mapValues { (_, v) -> v.makeInvariant() }
+  override fun get(index: String): TypeVariableName =
+    this@invariant[index].makeInvariant()
+}
+
 internal class ClassCompileScope(
   private val classDeclaration: KSClassDeclaration,
-  private val mutableCandidates: Sequence<KSClassDeclaration>,
+  private val mutableCandidates: Sequence<KSDeclaration>,
   override val logger: KSPLogger,
 ) : TypeCompileScope, KSClassDeclaration by classDeclaration {
 
-  override val typeVariableNames: List<TypeVariableName> = typeParameters.map { it.toTypeVariableName() }
+  override val typeVariableNames: List<TypeVariableName> =
+    classDeclaration.typeParameters.map { it.toTypeVariableName() }
+  override val typeParameterResolver: TypeParameterResolver
+    get() = classDeclaration.typeParameters.toTypeParameterResolver().invariant()
+
+  override val target: ClassName = classDeclaration.className
+  override val sealedTypes: Sequence<KSClassDeclaration> = classDeclaration.sealedTypes
+  override val properties: Sequence<KSPropertyDeclaration> = classDeclaration.getPrimaryConstructorProperties()
+
+  override val ClassName.parameterized
+    get() = parameterizedWhenNotEmpty(typeVariableNames.map { it.makeInvariant() })
+  override fun KSType.hasMutableCopy(): Boolean {
+    val closestDecl = declaration.closestClassDeclaration()
+    return closestDecl != null && closestDecl in mutableCandidates
+  }
+
+  override val KSPropertyDeclaration.typeName: TypeName
+    get() = type.toTypeName(typeParameterResolver).makeInvariant()
+
+  override fun toFileScope(file: FileSpec.Builder): FileCompilerScope = FileCompilerScope(this, file = file)
+}
+
+internal class TypeAliasCompileScope(
+  private val aliasDeclaration: KSTypeAlias,
+  private val mutableCandidates: Sequence<KSDeclaration>,
+  override val logger: KSPLogger,
+) : TypeCompileScope, KSTypeAlias by aliasDeclaration {
+
+  init {
+    require(aliasDeclaration.ultimateDeclaration != null)
```

Comment:
```suggestion
    requireNotNull(aliasDeclaration.ultimateDeclaration)
```

Also, I'm not sold on using init, but I guess it makes sense in this case